File: app/src/main/java/com/example/gradualalarm/MainActivity.java
package com.example.gradualalarm;


import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Intent;
import android.os.Bundle;
import android.provider.Settings;
import android.view.View;
import android.widget.TimePicker;
import android.widget.Toast;
import android.widget.ToggleButton;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import java.util.Calendar;

public class MainActivity extends AppCompatActivity {
    //App element used to pick time
    TimePicker alarmTimePicker;

    //Intent sent to a place
    PendingIntent pendingIntent;

    //class used to manage the alarm
    AlarmManager manager;

    String[] perms = new String[] {Settings.ACTION_REQUEST_SCHEDULE_EXACT_ALARM};

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        //TimePicker instantiation
        alarmTimePicker = (TimePicker) findViewById(R.id.timePicker);

        //AlarmManager instantiation
        manager = (AlarmManager) getSystemService(ALARM_SERVICE);

        //TODO: fix this its not requesting pemrissions
        if(!manager.canScheduleExactAlarms()){
            requestPermissions(perms,2);

        }

    }

    //Toggle alarm on/off
    public void OnToggleClicked(View view) {
        long time;
        //check if the toggle button is switched on

        if (((ToggleButton) view).isChecked()){

            //alarm on message displayed using Toast
            Toast.makeText(MainActivity.this, "ALARM ON", Toast.LENGTH_SHORT).show();
            Calendar calendar = Calendar.getInstance();

            //get calendar time from the TimePicker
            calendar.set(Calendar.HOUR_OF_DAY, alarmTimePicker.getHour());
            calendar.set(Calendar.MINUTE, alarmTimePicker.getMinute());

            //Intent set to AlarmReceiver which inherits from Broadcast Receiver
            Intent intent = new Intent(this, AlarmReceiver.class);

            //call the broadcast using pendingIntent
            pendingIntent = PendingIntent.getBroadcast(this, 0, intent, PendingIntent.FLAG_IMMUTABLE);

            //formula to round wake-up time to the exact minute
            time = (calendar.getTimeInMillis() - (calendar.getTimeInMillis() % 60000));

            //set the current time to AM or PM
            if (System.currentTimeMillis() > time){
                if (Calendar.AM_PM == 0)
                    time = time + (1000 * 60 * 60 * 12);
                else
                    time = time + (1000 * 60 * 60 * 24);
            }
            Toast.makeText(MainActivity.this, ("Time set to " + time), Toast.LENGTH_LONG).show();

            // Alarm rings continuously until toggle button is turned off
            manager.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, time, pendingIntent);
        }
        //turn alarm off otherwise
        else {
            manager.cancel(pendingIntent);

            //alarm off message
            Toast.makeText(MainActivity.this, "ALARM OFF", Toast.LENGTH_SHORT).show();
        }

    }
}

File: app/src/main/java/com/example/gradualalarm/MainActivity.kt
package com.geeksforgeeks.jctimepicker

import android.annotation.SuppressLint
import android.app.TimePickerDialog
import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.*
import androidx.compose.material.*
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.gradualalarm.AlarmManager
import java.util.*

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val alarmManager = AlarmManager(this)
        setContent {
            // Calling the composable function
            // to display element and its contents
            MainContent()
        }
    }
}

// Creating a composable
// function to display Top Bar
@SuppressLint("UnusedMaterial3ScaffoldPaddingParameter")
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun MainContent() {
    Scaffold(
        topBar = { TopAppBar(title = { Text("GFG | Time Picker", color = Color.Black) })},
        content = { MyContent() }
    )
}

fun setAlarm(){

}

@Composable
fun AlarmContent(){
    Column (
        modifier = Modifier.fillMaxSize(),
        horizontalAlignment =  Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center
    ){
        Button(onClick = {
            alarmManager.scheduleAlarm()
        }) {
            Text(text = "Set Alarm")
        }

        Spacer(modifier = Modifier.height(30.dp))
        Button(onClick = {
            alarmManager.cancelAlarm()
        }) {
            Text(text = "Cancel Alarm")
        }
    }
}

// Creating a composable function
// to create a Time Picker
// Calling this function as content
// in the above function
@Composable
fun MyContent(){

    // Fetching local context
    val mContext = LocalContext.current

    // Declaring and initializing a calendar
    val mCalendar = Calendar.getInstance()
    val mHour = mCalendar[Calendar.HOUR_OF_DAY]
    val mMinute = mCalendar[Calendar.MINUTE]

    // Value for storing time as a string
    val mTime = remember { mutableStateOf("") }

    // Creating a TimePicker dialog
    val mTimePickerDialog = TimePickerDialog(
        mContext,
        {_, mHour : Int, mMinute: Int ->
            mTime.value = "$mHour:$mMinute"
            //Todo: Insert function to set alarm time here
        }, mHour, mMinute, false
    )

    Column(modifier = Modifier.fillMaxSize(), verticalArrangement = Arrangement.Center, horizontalAlignment = Alignment.CenterHorizontally) {

        // On button click, TimePicker is
        // displayed, user can select a time
        Button(onClick = { mTimePickerDialog.show() }, colors = ButtonDefaults.buttonColors(Color(0XFF0F9D58))) {
            Text(text = "Open Time Picker", color = Color.White)
        }

        // Add a spacer of 100dp
        Spacer(modifier = Modifier.size(100.dp))

        // Display selected time
        Text(text = "Selected Time: ${mTime.value}", fontSize = 30.sp)
    }
}

// For displaying preview in
// the Android Studio IDE emulator
@Preview(showBackground = true)
@Composable
fun DefaultPreview() {
    MainContent()
}
